contextutil: add cancellation info log

madelynnblue opened the pull request:
We are currently attempting to track down various errors with context
cancelled messages (see #18652, #15332, #19199). In order to assist
in this debugging, add a contextutil package that wraps calls to
cancel with an info Log that prints the line that caused the cancel
to execute.

For now only selectively add this to suspect locations. Adding it to
all cancel funcs produces a huge amount of log spam.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19382)
<!-- Reviewable:end -->

RaduBerinde commented:
LGTM

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/util/contextutil/context.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382#-Kwq4jdGUj7FJxKJ0x7d:-Kwq4jdGUj7FJxKJ0x7e:b-eoz3dz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce40d09a966601285ab2aaa5ff106fd9fe478aa7/pkg/util/contextutil/context.go#L46)):*
> ```Go
> func wrap(ctx context.Context, cancel context.CancelFunc) (context.Context, context.CancelFunc) {
> 	return ctx, func() {
> 		log.InfofDepth(ctx, 1, "context canceled")
> ```

if V(1) maybe `debug.PrintStack()`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382#-:-Kwq4rQTUKLcYoqvl236:bb6ywmx)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/util/contextutil/context.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382#-Kwq4jdGUj7FJxKJ0x7d:-Kwq65scJnP2N3m0nYyh:b-bzw5sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce40d09a966601285ab2aaa5ff106fd9fe478aa7/pkg/util/contextutil/context.go#L46)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

if V(1) maybe `debug.PrintStack()`
</blockquote></details>

Done. Is this ok? I don't need to print the stack through the log package?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382)*
<!-- Sent from Reviewable.io -->

asubiotto commented on pkg/util/contextutil/context.go:
```diff
@@ -0,0 +1,53 @@
+
+func wrap(ctx context.Context, cancel context.CancelFunc) (context.Context, context.CancelFunc) {
+	return ctx, func() {
+		log.InfofDepth(ctx, 1, "context canceled")
```
Consider changing this line to `canceling context`, since `context canceled` is what's printed out as the `Err()` on a canceled context.

mjibson commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/contextutil/context.go, line 47 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382#-Kwq3Okf8LlxF7e8BIGa-r2-47:-Kwq7czafZbQcllNuYtp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0208aab63ad05d8c91787c8f3cc8c02d818a741a/pkg/util/contextutil/context.go#L47)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Consider changing this line to `canceling context`, since `context canceled` is what's printed out as the `Err()` on a canceled context.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/util/contextutil/context.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382#-Kwq4jdGUj7FJxKJ0x7d:-KwqBgh_4eJ_SUadHYAJ:bpc3cx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce40d09a966601285ab2aaa5ff106fd9fe478aa7/pkg/util/contextutil/context.go#L46)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Done. Is this ok? I don't need to print the stack through the log package?
</blockquote></details>

Good question.. Maybe `log.Infof(ctx, 1, "\n%s", debug.Stack())`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
I don't really know much about the locations you've chosen to instrument, but the idea and the `contextutil` code itself <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 5 of 6 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/util/contextutil/context.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382#-Kwq4jdGUj7FJxKJ0x7d:-KwqUAI6gnk4re7bn-49:b-eglj6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce40d09a966601285ab2aaa5ff106fd9fe478aa7/pkg/util/contextutil/context.go#L46)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Good question.. Maybe `log.Infof(ctx, 1, "\n%s", debug.Stack())`
</blockquote></details>

Passing `debug.Stack()` through the log package would be better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382#-:-KwqSoR-I3HYBiiyVzEx:b-tt9lhy)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/contextutil/context.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382#-Kwq4jdGUj7FJxKJ0x7d:-KwsYVSQNpQRB_GKTZp7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce40d09a966601285ab2aaa5ff106fd9fe478aa7/pkg/util/contextutil/context.go#L46)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Passing `debug.Stack()` through the log package would be better.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382)*
<!-- Sent from Reviewable.io -->

mjibson commented:
@knz Can you comment on the failing tests here? I'm not sure why they are happening.

madelynnblue committed:
contextutil: add cancellation info log

We are currently attempting to track down various errors with context
cancelled messages (see #18652, #15332, #19199). In order to assist
in this debugging, add a contextutil package that wraps calls to
cancel with an info Log that prints the line that caused the cancel
to execute.

For now only selectively add this to suspect locations. Adding it to
all cancel funcs produces a huge amount of log spam.

mjibson commented:
Updated to fix the show_trace with recommendations from @knz.

knz commented:
Reviewed 1 of 1 files at r5.
Review status: 6 of 7 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19382)*
<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/19382)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/util/contextutil/context.go, line 33 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/19382#-LL_l1RN4pzkZzH--QVn:-LL_l1RN4pzkZzH--QVo:b7xqcpm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c4f8ea8221ee6b79f4dc3ee23d27add3acc1736/pkg/util/contextutil/context.go#L33)):*
> ```Go
> 	return ctx, func() {
> 		if log.V(1) {
> 			log.InfofDepth(ctx, 1, "canceling context:\n%s", debug.Stack())
> ```

cc @jordanlewis 

Since I know there's work to rework this - note that logging something here, particularly a stack, is quite questionable since the returned `cancel` function is supposed to be always closed (generally after the ctx's work has finished). So in the common case it's just spam - it doesn't mean anyone will receive a `context.Canceled` error.
Ideally we'd have a way of distinguishing between canceling ongoing work with the intent to stop it and usual "cleanup" cancelation.


<!-- Sent from Reviewable.io -->
